src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,306 @@
+        public string ProcessFileName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the full path to the current project file.
+        /// </summary>
+        [Required]
+        public string ProjectFullPath { get; set; }
+
+        /// <summary>
+        /// Get or sets a value indicating whether or not the restore should restore all projects or just the entry project.
+        /// </summary>
+        public bool Recursive { get; set; }
+
+        /// <summary>
+        /// Gets or sets the full path to the solution file (if any) that is being built.
+        /// </summary>
+        public string SolutionPath { get; set; }
+
+        protected abstract string DebugEnvironmentVariableName { get; }
+
+        /// <inheritdoc cref="ICancelableTask.Cancel" />
+        public void Cancel() => _cancellationTokenSource.Cancel();
+
+        /// <inheritdoc cref="IDisposable.Dispose" />
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+
+            GC.SuppressFinalize(this);
+        }
+
+        public override bool Execute()
+        {
+            try
+            {
+#if DEBUG
+                var debugRestoreTask = Environment.GetEnvironmentVariable(DebugEnvironmentVariableName);
+                if (!string.IsNullOrEmpty(debugRestoreTask) && debugRestoreTask.Equals(bool.TrueString, StringComparison.OrdinalIgnoreCase))
+                {
+                    Debugger.Launch();
+                }
+#endif
+                DirectoryInfo tempDirectory = Directory.CreateDirectory(NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp));
+
+                FileInfo responseFile = new FileInfo(Path.Combine(tempDirectory.FullName, Path.GetRandomFileName()));
+
+                try
+                {
+                    WriteResponseFile(responseFile.FullName);
+
+                    MSBuildLogger logger = new MSBuildLogger(Log);
+
+                    using (var semaphore = new SemaphoreSlim(initialCount: 0, maxCount: 1))
+                    using (var loggingQueue = new TaskLoggingQueue(Log))
+                    using (var process = new Process())
+                    {
+                        process.EnableRaisingEvents = true;
+                        process.StartInfo = new ProcessStartInfo
+                        {
+                            Arguments = $"\"{string.Join("\" \"", GetCommandLineArguments(responseFile))}\"",
+                            CreateNoWindow = true,
+                            FileName = GetProcessFileName(ProcessFileName),
+                            RedirectStandardOutput = true,
+                            UseShellExecute = false,
+                            WorkingDirectory = Environment.CurrentDirectory,
+                        };
+
+                        // Place the output in the queue which handles logging messages coming through on StdOut
+                        process.OutputDataReceived += (sender, args) => loggingQueue.Enqueue(args?.Data);
```

Comment:
It looks like this is existing implementation, just refactored in this PR, I am guessing we should also add a handler for `Process.ErrorDataReceived` https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.errordatareceived?view=net-6.0 and log the errors the logger.

For example, 

https://github.com/NuGet/NuGet.Client/blob/451713c59f5206f77c97d093dd5191be13f9d0e0/test/TestUtilities/Test.Utility/CommandRunner.cs#L62-L101

This example has calls to `CancelOutputRead` and `CancelErrorRead` after process exits.

Reply:
The process we launch is `NuGet.Build.Tasks.Console.exe` which does not write anything to StdErr.  Instead, it logs JSON to StdOut (https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs#L149) and that tells the caller if a message is an error, warning, or message.